keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+    // This may be used for unlinking a symlink created through LinkVolume
+    rpc Rmdir(RmdirRequest) returns (RmdirResponse) {}
+
+    // StageDisk partitions a disk device and formats the partition with NTFS (if the disk has not been partitioned or formatted)
+    // The resulting volume is mounted at the requested global staging path.
+    rpc StageDisk(StageDiskRequest) returns (StageDiskResponse) {}
+
+    // StageSMBShare mounts a remote share over SMB on the host at the requested global staging path.
+    rpc StageSMBShare(StageSMBShareRequest) returns (StageSMBShareResponse) {}
+
+    // LinkVolume invokes mklink on the global staging path of a volume linking it to a path within a container
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskNumberWithLocation finds a disk device with requested Location <Bus, Target, LUN ID>
+    rpc GetDiskNumberWithLocation(GetDiskNumberWithLocationRequest) returns (GetDiskNumberWithLocationResponse) {}
+
+    // GetDiskNumberWithSCSIID finds a disk device whose SCSI Page 83 ID contains the passed ID string
+    rpc GetDiskNumberWithSCSIID(GetDiskNumberWithSCSIIDRequest) returns (GetDiskNumberWithSCSIIDResponse) {}
+}
+```
+
+In accordance with standard Kubernetes conventions, the above API will be introduces as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. New RPCs and enhancements to parameters will be introduced through v1alpha1 and graduate to v1beta1 and v1 versions as they mature.
+
+#### Enhancements in CSI Node Plugins
+
+Code for CSI Node Plugins need to be refactored to support CSI Node APIs in both Linux and Windows nodes. While the code targeting Linux nodes can assume privileged access to the host, the code targeting Windows nodes need to invoke the GRPC client API associated with the desired version of the CSIProxyService described above.
+
+Once compiled for Windows, container images based on Window Base images (like NanoServer) needs to be published and maintained.
+
+New YAMLs for DaemonSets associated with CSI Node Plugins needs to be authored that will (1) target Windows nodes and (2) use Windows paths for domain socket related paths as illustrated in the User Story section above.
+
+### Risks and Mitigations
+
+Any pod on a Windows node can be configured to mount `\\.\pipe\csi-proxy` and perform privileged operations. Thus csi-proxy presents a potential security risk. To mitigate the risk, an admission webhook will need to be implemented and deployed that will reject all containers that bind mount `\\.\pipe\csi-proxy` as a hostPath volume but does not have `privileged` flag set in the pod's securityContext specification. This allows the `privileged` setting to be used for Windows pods as an indication the container will perform privileged operations. Other cluster-wide policies (e.g. PSP) that act on the `privileged` setting in a container's `securityContext` can enforce the same for CSI Node plugins targeting Windows nodes.
```

Comment:
`Privileged` means a lot more than talking to the proxy on Linux. I wonder if there'd be concern for reusing this filed. 

/cc @tallclair

Reply:
I wanted to clarify something: in case of Linux, support for privileged containers already exist. So node plugins targeting Linux nodes will not require a proxy for any operations in the context of CSI node plugins. 

The scope of the proxy is thus purely Windows (and maybe expanded to other non Linux OS es later if Kubernetes expands support and those OS es lack privileged support). In case of Windows however, `Privileged` has no relevance today and therefore the proposal to use it.